File: iris-backend/app/schemas/ticket.py
"""Ticket schema."""
from marshmallow import Schema, fields, validate

PRIORITIES = ['Low', 'Medium', 'High', 'Critical']
STATUSES = ['Open', 'In Progress', 'Resolved', 'Closed']
CATEGORIES = ['Hardware Fault', 'Software Bug', 'Configuration', 'Training', 'Upgrade', 'Other']


class TicketSchema(Schema):
    ticket_id = fields.Int(dump_only=True)
    account_id = fields.Int(required=True)
    install_id = fields.Int(allow_none=True)
    ticket_ref = fields.Str(allow_none=True)
    title = fields.Str(required=True, validate=validate.Length(min=1, max=500))
    description = fields.Str(allow_none=True)
    priority = fields.Str(validate=validate.OneOf(PRIORITIES), load_default='Medium')
    status = fields.Str(validate=validate.OneOf(STATUSES), load_default='Open')
    category = fields.Str(allow_none=True, validate=validate.OneOf(CATEGORIES + [None]))
    raised_by = fields.Str(allow_none=True)
    raised_on = fields.DateTime(dump_only=True)
    resolved_on = fields.DateTime(allow_none=True)
    sla_breach = fields.Bool(load_default=False)
    assigned_to = fields.Int(allow_none=True)
    source = fields.Str(load_default='manual')
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)

File: iris-frontend/src/types/ticket.ts
export interface Ticket {
  ticket_id: number;
  account_id: number;
  account_name?: string;
  install_id?: number;
  product_name?: string;
  reference_number?: string;
  title: string;
  description?: string;
  status: 'Open' | 'In Progress' | 'Resolved' | 'Closed';
  priority: 'Critical' | 'High' | 'Medium' | 'Low';
  category?: string;
  source: 'manual' | 'api';
  raised_by?: string;
  assigned_to?: string;
  sla_breach: boolean;
  raised_on: string;
  resolved_on?: string;
  created_at: string;
  updated_at: string;
}

export interface TicketFilters {
  search?: string;
  account_id?: number;
  status?: string;
  priority?: string;
  category?: string;
  date_from?: string;
  date_to?: string;
  page?: number;
  per_page?: number;
}

export interface CreateTicketPayload {
  account_id: number;
  install_id?: number;
  title: string;
  description?: string;
  priority: 'Critical' | 'High' | 'Medium' | 'Low';
  category?: string;
  raised_by?: string;
}
